### PR TITLE
Add documentation and example of arbitrary QAT flow used with recipe-based export.

### PR DESCRIPTION
**Pull request by MartinPavella**
### Summary

Adds a runnable example and documentation showing how to run an arbitrary Quantization-Aware Training flow in the middle of the declarative export recipe, by splitting the recipe around the `QUANTIZE` stage.

The declarative `ExportRecipe` / `ExportSession` path runs the whole pipeline as one `export()` call, which is fine for simple calibration but cannot host complex QAT — real training runs involve iterative loops, checkpointing, job restarts, validation-accuracy evaluation, rollback on regression, and possibly separate machines. The recipe's `pipeline_stages` attribute is used to run the pre-quantize stages, hand the graph off to an out-of-band QAT flow as a `.pt2`, and resume the recipe from `TORCH_EXPORT` onward.

1. New example under `examples/export/qat_pipeline_split/`: four independently-invokable stages driven by `run.sh`. `1_prepare.py` captures up to the quantize boundary, `2_qat.py` prepares with `prepare_qat_pt2e` and does a stand-in train + checkpoint round-trip before `convert_pt2e`, `3_lower.py` lowers to `model.pte` via the recipe, and `4_run.py` runs the `.pte` through the runtime. Each stage asserts its own invariants, so the example doubles as a smoke test.

2. The example shows one method with optional simplifications: the full split sets `pipeline_stages` explicitly on both halves, while the simplified path skips the pre-quantize slice and lets the recipe auto-skip `SOURCE_TRANSFORM` / `QUANTIZE` / `TORCH_EXPORT` for `ExportedProgram` input (default pipeline only).

3. New doc `docs/source/using-complex-qat-flows-with-recipe-based-lowering.md`, registered in the `usage.md` toctree and cross-linked from `quantization-overview.md`. `using-executorch-export.md` was left untouched since it documents only the imperative API.

Authored with Vero Code.

cc @mergennachin @nil-is-all @iseeyuan @lucylq @helunwencser @tarun292 @kimishpatel @jackzhxng @jerryzh168 @metascroy

### CODE REVIEW

File: examples/export/qat_pipeline_split/1_prepare.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Stage 1: capture the model up to the quantize boundary.
+#
+# Two example paths (selected with --example):
+#
+#   minimal (default)
+#       Export the eager nn.Module to an ATEN-dialect ExportedProgram with a
+#       single torch.export.export call.  No recipe, no ExportSession.
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/1_prepare.py]: nit: double space in the comment

File: examples/export/qat_pipeline_split/1_prepare.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+#
+#   sliced
+#       Drive only the SOURCE_TRANSFORM stage through ExportSession, then
+#       export to ATEN and save.  This shows how to restrict the pipeline to
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/1_prepare.py]: nit: double space in the comment

File: examples/export/qat_pipeline_split/model.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# Copyright 2026 NXP
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Model definition for the QAT pipeline split example.
+#
+# SmallConvNet is used only by stage 1 (1_prepare.py), which constructs the
+# eager model before capture.  Stages 2 and 3 reload the model from a saved
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/model.py]: nit: double space in the comment

File: examples/export/qat_pipeline_split/1_prepare.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    ex = get_example_inputs()
+
+    assert isinstance(
+        model, torch.nn.Module
+    ), f"Expected nn.Module from get_model(), got {type(model)}"
+
+    ep = torch.export.export(model, ex, strict=True)
+
+    assert ep is not None, "torch.export.export returned None"
+    print(f"[minimal] Graph nodes: {len(list(ep.graph.nodes))}")
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/1_prepare.py:51]: nit: `Graph nodes` -> `Number of graph nodes` is better imo, however I wouldn't mind if the whole list of nodes was printed.

File: examples/export/qat_pipeline_split/3_lower.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+    pt2_in = os.path.join(workdir, "stage2_minimal_quantized.pt2")
+    assert os.path.isfile(pt2_in), (
+        f"Input file not found: {pt2_in}  " "(run 2_qat.py --example minimal first)"
+    )
+    print(f"[minimal] Lowering quantized model from {pt2_in}")
+
+    recipe = _build_recipe()
+
+    # Pass the file path directly.  ExportSession auto-skips SOURCE_TRANSFORM,
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/3_lower.py]: nit: again some weird double space. Please fix in all occurrences across the changes.

File: examples/export/qat_pipeline_split/3_lower.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    print(f"\n  Saved -> {pte_path}  ({os.path.getsize(pte_path)} bytes)")
+
+
+def run_minimal(workdir: str) -> None:
+    from executorch.export import export as et_export
+    from executorch.export.types import StageType
+
+    pt2_in = os.path.join(workdir, "stage2_minimal_quantized.pt2")
+    assert os.path.isfile(pt2_in), (
+        f"Input file not found: {pt2_in}  " "(run 2_qat.py --example minimal first)"
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/3_lower.py]: nit: I'd replace `minimal` in `...run 2_qat.py --example minimal first...` with `{workdir}`

File: docs/source/using-complex-qat-flows-with-recipe-based-lowering.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+### Post-quantize slice
+
+Continue the recipe pipeline from `TORCH_EXPORT` onward:
+
+```python
+recipe.pipeline_stages = [
+    StageType.TORCH_EXPORT,
+    StageType.TO_EDGE_TRANSFORM_AND_LOWER,
```
Comment (JakeStevens) [inline docs/source/using-complex-qat-flows-with-recipe-based-lowering.md:119]: what about

`EDGE_PROGRAM_MANAGER_TRANSFORM`

File: examples/export/qat_pipeline_split/2_qat.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+from model import get_calibration_inputs, get_example_inputs
+
+
+def _get_quantizer():
+    from executorch.backends.xnnpack.recipes.xnnpack_recipe_types import (
+        XNNPackRecipeType,
+    )
+    from executorch.export.recipe import ExportRecipe
+
+    recipe = ExportRecipe.get_recipe(XNNPackRecipeType.PT2E_INT8_STATIC_PER_TENSOR)
```
Comment (JakeStevens) [inline examples/export/qat_pipeline_split/2_qat.py:30]: fwiw this recipe sets qat = false

File: examples/export/qat_pipeline_split/README.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+every stage of the recipe with no overlap and no stage silently skipped.
+
+Use this when you need pre-quantize recipe passes to run before QAT (for
+example, torchao source transforms that the recipe provides).
+
+### Which should I use?
+
+Use **`minimal`** when the recipe uses its default pipeline stages and you have
+no need for pre-quantize recipe passes.  The `minimal` mode works because
+`ExportSession` auto-skips SOURCE_TRANSFORM, QUANTIZE and TORCH_EXPORT when it
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/README.md]: nit: I'd put SOURCE_TRANSFORM, QUANTIZE and TORCH_EXPORT to code brackets, like `SOURCE_TRANSFORM`...

File: examples/export/qat_pipeline_split/README.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+PYTHON=/path/to/python bash examples/export/qat_pipeline_split/run.sh \
+    --example sliced \
+    --workdir /tmp/qat_pipeline_split_sliced
+```
+
+Each stage can also be run independently:
+
+```bash
+python 1_prepare.py --example minimal --workdir /tmp/out
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/README.md:141]: nit: I'd change `python` to `PYTHON=/path/to/python` to keep it consistent with the lines above.

File: examples/export/qat_pipeline_split/2_qat.py (hunk 8 above)
Comment (novak-vaclav) [reply examples/export/qat_pipeline_split/2_qat.py]: blocking: I agree with Jake's comment regarding the QAT being `False` by default here

File: examples/export/qat_pipeline_split/3_lower.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        quantized_gm, torch.fx.GraphModule
+    ), f"Expected GraphModule after load, got {type(quantized_gm)}"
+    print(
+        f"[sliced]  Loaded GraphModule ({len(list(quantized_gm.graph.nodes))} nodes)."
+    )
+
+    recipe = _build_recipe()
+
+    # Mirror image of the [SOURCE_TRANSFORM] slice in stage 1: together they
+    # cover the full pipeline with no overlap and no stage silently skipped.
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/3_lower.py]: suggestion: `QUANTIZE` stage with mode `sliced` is indeed silently skipped. I'd add this note to places where `sliced` and no-skipping behavior is mentioned to avoid confusion.

File: examples/export/qat_pipeline_split/4_run.py [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    args = parser.parse_args()
+
+    pte_path = os.path.join(args.workdir, "model.pte")
+    assert os.path.isfile(
+        pte_path
+    ), f"model.pte not found at {pte_path}  (run 3_lower.py first)"
+
+    try:
+        from executorch.runtime import Runtime
+    except ModuleNotFoundError:
```
Comment (novak-vaclav) [inline examples/export/qat_pipeline_split/4_run.py]: nit: I'd catch `ModuleNotFoundError` as it is broader exception

**Review by novak-vaclav — COMMENTED** [review]
From me, there is one blocking comment, one suggestion, and a couple of nits. Otherwise good job 😊 I think the whole concept is explained really well

Have you thought about adding some tests for this? There are a bunch of hardcoded strings, or a hardcoded shape of ET runtime output. If any of this silently changes, the example may start failing unexpectedly.

File: examples/export/qat_pipeline_split/4_run.py (hunk 12 above)
Comment (MartinPavella) [reply examples/export/qat_pipeline_split/4_run.py]: I assume you meant `ImportError`. I will use that instead.

File: examples/export/qat_pipeline_split/2_qat.py (hunk 8 above)
Comment (MartinPavella) [reply examples/export/qat_pipeline_split/2_qat.py]: As the quantization stage is skipped and handled separately, this doesn't really matter as long as the custom quantization flow uses a QAT quantizer (which the example wasn't, so I have fixed it now, thanks).

File: docs/source/using-complex-qat-flows-with-recipe-based-lowering.md (hunk 7 above)
Comment (MartinPavella) [reply docs/source/using-complex-qat-flows-with-recipe-based-lowering.md:119]: The `XNNPACK` recipe doesn't use `edge_manager_transform_passes`, so the default pipeline doesn't contain the `EDGE_PROGRAM_MANAGER_TRANSFORM` stage.

I have added a comment stating that other recipes may need that stage though.

File: examples/export/qat_pipeline_split/README.md (hunk 10 above)
Comment (MartinPavella) [reply examples/export/qat_pipeline_split/README.md:141]: The lines above export the variable for use inside the bash scripts. Here, that is not necessary. I will keep it as is.

File: examples/export/qat_pipeline_split/4_run.py (hunk 12 above)
Comment (novak-vaclav) [reply examples/export/qat_pipeline_split/4_run.py]: Yes, that's what I meant, sorry 😂

**Review by novak-vaclav — APPROVED** [review]
Good job, LGTM 😊👌

**Review by robert-kalmar — APPROVED** [review]

**Comment by robert-kalmar** [comment]
The example looks good to me, it shows how to use the recipes with the QAT workflow. Additionally (not in this PR) we shall add documentation to recipes itself too. At this moment the there is almost no notion about recipes. 

Good work @MartinPavella . 

@JakeStevens , @rascani, @mergennachin , how you see the documentation and the example flow? Any objectives?
